ark/internal/controller/query_controller.go

```diff
@@ -678,16 +701,18 @@ func (r *QueryReconciler) executeModel(ctx context.Context, query arkv1alpha1.Qu
 
 	responseMessages := []genai.Message{assistantMessage}
 
-	// Save new messages to memory (user message + response messages)
-	newMessages := append([]genai.Message{userMessage}, responseMessages...)
+	// Save all new messages (input + response) to memory
+	newMessages := make([]genai.Message, 0, len(inputMessages)+len(responseMessages))
+	newMessages = append(newMessages, inputMessages...)
+	newMessages = append(newMessages, responseMessages...)
 	if err := memory.AddMessages(ctx, query.Name, newMessages); err != nil {
 		return nil, fmt.Errorf("failed to save new messages to memory: %w", err)
 	}
 
 	return responseMessages, nil
 }
 
-func (r *QueryReconciler) executeTool(ctx context.Context, query arkv1alpha1.Query, toolName string, impersonatedClient client.Client, tokenCollector *genai.TokenUsageCollector) ([]genai.Message, error) { //nolint:unparam
+func (r *QueryReconciler) executeTool(ctx context.Context, query arkv1alpha1.Query, inputMessages []genai.Message, toolName string, impersonatedClient client.Client, tokenCollector *genai.TokenUsageCollector) ([]genai.Message, error) { //nolint:unparam
```

Comment:
should we ever do this or should content of type `messages` never go to tools, cause it's pretty much by definition wrong